add flux jobtap query subcommand

grondo opened the pull request:
As described in #4506, this PR adds support for a new `flux jobtap query` subcommand which can fetch extended data about jobtap plugins (currently `name` and `path` by default). Optionally, plugins can register for a `plugin.query` callback topic and add arbitrary data to the response payload. This gives plugins a generic way to export internal state for debugging or other purposes.

As an example implementation, the `dependency-after` plugin is extended to support this interface and adds a `dependencies` key the query response which includes all active dependencies, e.g.:

```
$ flux jobtap query .dependency-after | jq
{
  "name": ".dependency-after",
  "path": "builtin",
  "dependencies": [
    {
      "id": 500699234304,
      "depid": 505581404160,
      "type": "after-success",
      "description": "after-success=fE9r5TG3"
    },
    {
      "id": 519590379520,
      "depid": 524053118976,
      "type": "after-success",
      "description": "after-success=fEedRWaK"
    },
    {
      "id": 533515468800,
      "depid": 537894322176,
      "type": "after-success",
      "description": "after-success=fF1qwEUX"
    }
  ]
}
```

If a plugin was loaded from a file, the path reflects the actual path of the original DSO:

```
$ flux jobtap query alloc-bypass.so | jq
{
  "name": "alloc-bypass.so",
  "path": "/home/grondo/git/flux-core.git/src/modules/job-manager/plugins/.libs/alloc-bypass.so"
}
```

cmoussa1 reviewed (approved):
This LGTM! Thanks @grondo for working on this. I think this will be a useful utility especially for the multi-factor priority plugin. I also think it should be somewhat straightforward for me to adjust flux-framework/flux-accounting#264 as a result of this PR, which I will do once this lands.

Thanks!

grondo commented:
@garlick noticed some `errno` inconsistencies in this PR and I've just forced a push to address those (largely by not using `errno` anymore)
 - The query RPC no longer sets the `errno` parameter to `flux_respond_error(3)` since I think in all cases a human-readable error is used
 -  `flux_plugin_get_path(3)` no longer sets EINVAL when its parameter is `NULL`

I also fixed a few tests that failed to supply the `jq` `-e` option, and added a couple tests for good measure.

garlick commented on src/modules/job-manager/jobtap.c:
```diff
@@ -1425,6 +1425,93 @@ void jobtap_handler (flux_t *h,
+        goto error;
+    }
+
+    rc = flux_respond (h, msg, result);
+    ERRNO_SAFE_WRAP (free, result);
+    if (rc < 0)
+        goto error;
+    return;
```
If `flux_respond()` fails, best to just log the error and return since the error response will likely fail too.

grondo replied:
Duh, good point.

garlick commented on src/modules/job-manager/plugins/dependency-after.c:
```diff
@@ -617,6 +623,63 @@ static int inactive_cb (flux_plugin_t *p,
+        return -1;
+    }
+
+    if (flux_plugin_arg_pack (args,
+                              FLUX_PLUGIN_ARG_OUT,
+                              "{s:o}",
+                              "dependencies",
+                              o) < 0)
+        flux_log_error (flux_jobtap_get_flux (p),
+                        "dependency-after: query_cb: flux_plugin_arg_pack: %s",
+                        flux_plugin_arg_strerror (args));
```
Leaks `o` on `flux_plugin_arg_pack()` failure.

garlick commented on doc/man7/flux-jobtap-plugins.rst:
```diff
@@ -237,6 +237,21 @@ The callback should return 0 on success, and -1 on failure.  On failure,
+  used by a plugin to export internal plugin state for inspection by an admin
+  or user by placing the data in the output arguments of the callback, e.g.::
+
+    flux_plugin_arg_pack (p, FLUX_PLUGIN_ARG_OUT,
+                          "{s:o}"
+                          "data", internal_data);
```
Since the args are destroyed after the callback returns, maybe the example should use `{s:O}` (big O)?

I'm not sure if that makes it any clearer.  This is a fertile ground for bugs though.

grondo replied:
Ah, very good point. Given my mistake above, it is probably always better to pack with "O" and decref manually. (I don't know how many times I have to relearn this until it gets through my head)

garlick reviewed (approved):
Nicely done!  Just a couple of nits.

grondo commented:
I addressed @garlick's comments above and force-pushed the result.

garlick commented:
Thanks - feel free to set MWP!

grondo commented:
Thanks!
